src/client/Microsoft.Identity.Client/Internal/Requests/ClientCredentialRequest.cs

```diff
@@ -153,6 +162,35 @@ private async Task<MsalTokenResponse> SendTokenRequestToProviderAsync(Cancellati
             return tokenResponse;
         }
 
+        private async Task<MsalAccessTokenCacheItem> TryGetCachedAccessTokenAsync()
+        {
+            MsalAccessTokenCacheItem cachedAccessTokenItem = await CacheManager.FindAccessTokenAsync().ConfigureAwait(false);
+
+            if (cachedAccessTokenItem != null && !_clientParameters.ForceRefresh)
```

Comment:
so the first cache check is done in line 56, and when cache is empty then we sent the call to IDP. In a multi threaded scenario, the calls would have already gone past line 56 and waiting to hit the IDP call. now we only allow one thread to execute the IDP call, fill the cache. The next thread that already passed 56, will now execute the IDP call, before it does we are checking cache one more time, so the cache that was filed by the first thread can be used,.

Reply:
> does we are checking cache one more time, so the cache that was filed by the first thread can be used,.

Yea, and my point is that if ForceRefresh is true, we don't need to check the cache, just go to IDP

What we have now in this method (if cached token is found):
- Check cache
- If ForceRefresh > go to IDP.
- If !ForceRefresh > return found token.

Proposed:
- If ForceRefresh > go to IDP.
- If !ForceRefresh > check cache, then return found token or go to IDP.

Reply:
This could also be changed to were the parent method checks ForceRefresh flag and calls GetFromCache if it's false

Reply:
> This could also be changed to were the parent method checks ForceRefresh flag and calls GetCachedAT if it's false. So GetCachedAT doesn't need extra ForceRefresh check.

Yes, now I get it because I did not change that logic. Let me modify it.